Realm/ObjectServerTests/SwiftObjectServerTests.swift

```diff
@@ -596,39 +907,353 @@ class SwiftObjectServerTests: SwiftSyncTestCase {
+        var config = user.flexibleSyncConfiguration(clientResetMode: .recoverOrDiscardUnsyncedChanges(assertBeforeBlock, assertAfterBlock))
+        config.objectTypes = [SwiftPerson.self]
+        guard let syncConfig = config.syncConfiguration else {
+            fatalError("Test condition failure. SyncConfiguration not set.")
+        }
+        switch syncConfig.clientResetMode {
+        case .recoverOrDiscardUnsyncedChanges(let before, let after):
+            XCTAssertNotNil(before)
+            XCTAssertNotNil(after)
+        default:
+            XCTFail("Should be set to recoverOrDiscard")
+        }
+
+        try autoreleasepool {
+            XCTAssertEqual(user.flexibleSyncConfiguration().fileURL, config.fileURL)
+            let realm = try Realm(configuration: config)
+            let subscriptions = realm.subscriptions
+            XCTAssertEqual(subscriptions.count, 1) // subscription created during prepareFlexibleSyncClientReset
+            XCTAssertEqual(subscriptions.first?.name, "all_people")
+
+            waitForExpectations(timeout: 15.0)
+            XCTAssertEqual(realm.objects(SwiftPerson.self).count, 1)
+            // The Person created locally ("John") should have been discarded,
+            // while the one from the server ("Paul") should be present.
+            XCTAssertEqual(realm.objects(SwiftPerson.self)[0].firstName, "Paul")
+        }
+        try waitForEditRecoveryMode(flexibleSync: true, appId: self.flexibleSyncAppId, disable: false)
+    }
+
+    // A stray session is left during teardown. Can't find out where.
+    func fix_testFlexibleClientResetManual() throws {
+        try autoreleasepool {
+            let user = try logInUser(for: basicCredentials(app: self.flexibleSyncApp), app: self.flexibleSyncApp)
+            try prepareFlexibleClientReset(user)
+
+            var config = user.flexibleSyncConfiguration(clientResetMode: .manual(assertManualClientReset(user)))
+            config.objectTypes = [SwiftPerson.self]
+
+            guard let syncConfig = config.syncConfiguration else {
+                fatalError("Test condition failure. SyncConfiguration not set.")
+            }
+            switch syncConfig.clientResetMode {
+            case .manual(let block):
+                XCTAssertNotNil(block)
+            default:
+                XCTFail("Should be set to manual")
+            }
+            try autoreleasepool {
+                let realm = try Realm(configuration: config)
+                waitForExpectations(timeout: 15.0)
+                // The locally created object should still be present as we didn't
+                // actually handle the client reset
+                XCTAssertEqual(realm.objects(SwiftPerson.self).count, 1)
+                XCTAssertEqual(realm.objects(SwiftPerson.self)[0].firstName, "John")
+            }
+//            resetSyncManager() Need to reset the syncManager. But path_for_realm is nil when creating new flex config?
```

Comment:
The session is torn down asynchronously after the last Realm instance is released. `syncManager.waitForSessionTermination()` waits for all sessions to finish being destroyed.